### PR TITLE
[Bugfix][Step-Audio2] Align audio feature lengths with encoder outputs

### PR DESCRIPTION
**Pull request by LHXuuu**
## Background

Step-Audio2 applies 479 samples of right padding before computing the log-mel spectrogram. As a result, even a very short non-empty audio tail can produce three mel frames.

However, `padding_mels` previously reported each mel length as `T - 2`, excluding the two boundary frames. For a one-sample audio tail:

- The actual mel length is `3`.
- The reported mel length becomes `1`.
- Prompt expansion forces at least one audio placeholder.
- The model encoder calculates zero valid audio embeddings from the reported length.

This causes the number of multimodal embeddings to differ from the number of placeholders. A real request containing ten 25-second chunks followed by a very short tail produced:

```text
3130 multimodal embeddings
3131 audio placeholders
```

The request then failed while merging multimodal embeddings into the language-model inputs.

## Purpose

Ensure that Step-Audio2 uses the same audio-length convention for prompt expansion, encoder masking, and adapter output slicing.

In particular, a non-empty three-frame audio tail should produce exactly one placeholder and one embedding instead of one placeholder and zero embeddings.

## Changes

- Preserve the full mel-spectrogram length in `padding_mels`, including the two boundary frames introduced by Step-Audio2's right padding.
- Calculate the encoder output length using the same two-stage downsampling formula as the model:

  ```python
  (mel_length + 1) // 2 // 2
  ```

- Remove the forced minimum of one feature from `calculate_audio_feature_length`, so the helper reports the actual encoder and adapter output length.
- Add regression coverage for:
  - A one-sample audio tail producing three mel frames.
  - The boundary mappings `3 -> 1` and `2 -> 0`.
  - Ten full 25-second chunks plus a short tail producing:

    ```text
    [313] * 10 + [1] = 3131
    ```

  - The encoder and adapter producing one valid embedding for a three-frame input.

## Test Plan

Run the targeted Step-Audio2 unit tests:

```bash
pytest -q tests/model_executor/models/step_audio2/test_step_audio2_thinker.py
```

Run end-to-end ASR requests covering:

1. A standalone 10 ms audio input.
2. Ten 25-second audio chunks followed by a 10 ms tail.
3. A real speech sample followed by a 10 ms tail to verify that ASR quality is preserved.

**vLLM Version:** `0.25.0`

**vLLM-Omni Commit:** `253740c4124d361ea91471f5bed85a4ca80ad359`

### Short-tail boundary test

Command:

```bash
curl --noproxy '*' -sS \
  -X POST 'http://127.0.0.1:8188/v1/chat/completions' \
  -H 'Content-Type: application/json' \
  --data-binary @step_audio_10ms_request.json \
  -w '\nHTTP %{http_code}\n'
```

The request contains a 10 ms, 16 kHz audio input. It produces three mel frames and one audio embedding.

Result:

```text
HTTP 200
finish_reason: stop
prompt_tokens: 32
completion_tokens: 1
```

The previous `0 multimodal tokens to 1 placeholders` error no longer occurs.

### Original failure regression test

Command:

```bash
curl --noproxy '*' -sS \
  -X POST 'http://127.0.0.1:8188/v1/chat/completions' \
  -H 'Content-Type: application/json' \
  --data-binary @step_audio_10x25s_plus_tail_request.json \
  -w '\nHTTP %{http_code}\n'
```

The request contains ten 25-second audio chunks followed by a 10 ms tail.

Expected feature lengths:

```text
[313] * 10 + [1] = 3131
```

Result:

```text
HTTP 200
prompt_tokens: 3162
completion_tokens: 16
finish_reason: length
```

The previous `3130 multimodal tokens to 3131 placeholders` error no longer occurs.

### ASR correctness regression test

Command:

```bash
curl --noproxy '*' -sS \
  -X POST 'http://127.0.0.1:8188/v1/chat/completions' \
  -H 'Content-Type: application/json' \
  --data-binary @step_audio_mary_had_lamb_with_tail_request.json \
  -w '\nHTTP %{http_code}\n'
```

The request contains the vLLM `mary_had_lamb.ogg` sample followed by a 10 ms audio tail.

Result:

```text
HTTP 200
prompt_tokens: 263
completion_tokens: 46
finish_reason: stop
```

Transcription:

```text
The first words I spoke in the original kornograph, a little piece of
practical poetry: Mary had a little lamb, its fleece was white as snow,
and everywhere that Mary went, the lamb was sure to go.
```

The spoken content was transcribed correctly apart from one minor word substitution, and the 10 ms tail did not introduce additional text.

```

### CODE REVIEW

**Comment by yuanheng-zhao** [comment]
Is this PR resolving the bug reported by #5434 , or any related issue?

**Comment by yuanheng-zhao** [comment]
> Is this PR resolving the bug reported by #5434 , or any related issue?

Seems not the same bug. Shall we create and link an issue recording the reproduction steps

File: vllm_omni/model_executor/models/step_audio2/step_audio2_thinker.py [hunk 1]
```diff
@@ -344,11 +351,11 @@ def calculate_audio_feature_length(mel_length: int) -> int:
         >>> calculate_audio_feature_length(1000)
         125
     """
-    encoder_output_len = (mel_length + 1) // 4
+    encoder_output_len = (mel_length + 1) // 2 // 2
```
Comment (yuanheng-zhao) [inline vllm_omni/model_executor/models/step_audio2/step_audio2_thinker.py:354]: nit: This seems to be a no-op

File: vllm_omni/model_executor/models/step_audio2/step_audio2_thinker.py [hunk 2]
```diff
@@ -117,7 +117,14 @@ def padding_mels(data: list[torch.Tensor]) -> tuple[torch.Tensor, torch.Tensor]:
...
     """
-    feats_lengths = torch.tensor([s.size(1) - 2 for s in data], dtype=torch.int32)
+    # Keep the full mel length, including the two boundary frames produced by
+    # Step-Audio2's 479-sample right padding.  The Step-Audio2 reference
+    # implementation uses this same length for prompt expansion and encoder
+    # masking.  Dropping the boundary frames here makes very short, non-empty
+    # audio inconsistent:
+    # prompt expansion produces one placeholder while the encoder produces no
+    # embeddings.
+    feats_lengths = torch.tensor([s.size(1) for s in data], dtype=torch.int32)
```
Comment (yuanheng-zhao) [inline vllm_omni/model_executor/models/step_audio2/step_audio2_thinker.py:127]: Is the removal of minusing two boundary frames making it inconsistent with upstream? I.e., two paths of returned `feats_lengths` and `calculate_audio_feature_length` inconsistent, https://github.com/stepfun-ai/Step-Audio2/blob/ea949018fa5718f7fa4e539dd46f2aee04bfbf26/utils.py#L56-L69

File: tests/model_executor/models/step_audio2/test_step_audio2_thinker.py [hunk 3]
```diff
@@ -123,6 +153,14 @@ def test_audio_encoder_output_shape():
     assert adapted.shape[0] == batch_size
     assert adapted.shape[2] == 4096
 
+    short_tail = torch.randn(1, n_mels, 3)
+    short_encoded, short_encoded_lens = encoder(short_tail, torch.tensor([3], dtype=torch.int32))
+    short_adapted = adapter(short_encoded)
+    short_feature_lens = (short_encoded_lens - 1) // 2 + 1
+
+    assert short_feature_lens.tolist() == [1]
+    assert short_adapted.shape[1] >= short_feature_lens.item()
```
Comment (yuanheng-zhao) [inline tests/model_executor/models/step_audio2/test_step_audio2_thinker.py:162]: why this assertion is `>=` rather than checking equality

File: vllm_omni/model_executor/models/step_audio2/step_audio2_thinker.py (hunk 2 above)
Comment (linyueqian) [inline vllm_omni/model_executor/models/step_audio2/step_audio2_thinker.py:127]: [important] Reporting the full mel length changes placeholder counts well beyond the short-tail case this PR targets, and it changes what the encoder attends to on every request.

On counts: the formula in isolation only differs for a mel length of 2 or less, but the value fed into it also changed here, from `T - 2` to `T`. Comparing the actual pipeline, old `calculate_audio_feature_length(T - 2)` against new `calculate_audio_feature_length(T)`, the count differs by one for every `T` congruent to 3 or 4 modulo 8 above 4. `T = 11` and `T = 12` both go from 1 placeholder to 2, and that pattern repeats every eight frames for the whole range. The two values the tests pin, 2502 and 3, are among the ones where old and new happen to agree, so the suite does not cover any length where this change is visible.

On the encoder: `x_len` is not just bookkeeping. `AudioEncoder.forward` builds its attention mask with `make_non_pad_mask(x_len, T)`, so raising the reported length from `T - 2` to `T` unmasks the two STFT boundary frames produced by the 479-sample right padding. Attention is global, so every output position can shift, for all audio and not only for short tails. The output width is unchanged on the 25 second path because the encoder applies the same `(x_len + 1) // 2 // 2` arithmetic, which is also why placeholders and embeddings stay consistent and the reported crash does stop.

So the fix does remove the mismatch. What I cannot confirm is that it removes it in the direction the model expects, which is the subject of the question below.

File: vllm_omni/model_executor/models/step_audio2/step_audio2_thinker.py [hunk 4]
```diff
@@ -117,7 +117,14 @@ def padding_mels(data: list[torch.Tensor]) -> tuple[torch.Tensor, torch.Tensor]:
         padded_feats: Padded tensor of shape (batch, n_mels, max_T)
         feats_lengths: Lengths of each sequence
     """
-    feats_lengths = torch.tensor([s.size(1) - 2 for s in data], dtype=torch.int32)
+    # Keep the full mel length, including the two boundary frames produced by
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/step_audio2/step_audio2_thinker.py:120]: [question] This comment carries the entire justification for the change, and I cannot verify it from this repository. Nothing in this checkout vendors the Step-Audio2 reference length convention, and there is no `compute_token_num` or equivalent to compare against, so the claim that the reference uses the full mel length for prompt expansion and encoder masking is unverifiable here.

The distinction matters because the encoder weights are pretrained. If the reference masks at `T - 2`, then this change feeds a frozen encoder two padding-derived frames it was trained to ignore, and the alternative minimal fix is to keep the `T - 2` reporting, drop the `max(1, ...)` clamp so a very short tail yields 0 rather than a forced 1, and prevent the `default_feature_len` fallback from firing when the reported length is 0. That also makes placeholders and embeddings agree, without changing the encoder's input convention or the counts at `T` congruent to 3 or 4 modulo 8.

Could you point at the specific upstream lines that use the full length? If they say what the comment says, this is right as written and I will approve it. If they use `T - 2`, the smaller fix looks preferable.

File: tests/model_executor/models/step_audio2/test_step_audio2_thinker.py [hunk 5]
```diff
@@ -44,11 +44,29 @@ def test_audio_preprocessing():
...
+    )
+
+    full_chunk = torch.zeros(128, 2502)
+    short_tail = log_mel_spectrogram(np.zeros(1, dtype=np.float32))
+
+    padded_mels, mel_lens = padding_mels([full_chunk, short_tail])
+
+    assert short_tail.shape == (128, 3)
+    assert padded_mels.shape == (2, 128, 2502)
+    assert mel_lens.tolist() == [2502, 3]
```
Comment (linyueqian) [inline tests/model_executor/models/step_audio2/test_step_audio2_thinker.py:69]: [suggestion] These assertions pin the two lengths where the old and new conventions agree, 2502 and 3, so they lock in the new convention without exercising any case that distinguishes it. A case at a length congruent to 3 modulo 8, `T = 11` for example, would cover where the count actually moves, from 1 placeholder under the old reporting to 2 under the new one. That is the assertion a future reader needs in order to see what this change decided.

**Review by linyueqian — COMMENTED** [review]
Reviewed at 253740c4. The diagnosis is right and the mechanism is real: padding_mels reported T minus 2 while the encoder consumed the full padded tensor, so prompt expansion and embedding generation disagreed, and a short tail produced the 3131 against 3130 in the description. The patch does make the two sides agree, because the encoder applies the same halving arithmetic to whatever length it is given.

My question is about direction rather than consistency, and it is inline. Changing the reported length also changes which frames the encoder attends to, since AudioEncoder.forward derives its attention mask from that length, so the two boundary frames created by the 479-sample right padding become visible to a pretrained encoder on every request. It also moves the placeholder count for a quarter of possible mel lengths, not only for the very short tails, which the current tests do not cover because they pin the two lengths where both conventions agree. The comment introducing the change says this matches the Step-Audio2 reference, and that is the load-bearing claim, but nothing in this checkout vendors the reference for me to check it against.

If the reference does report the full length, this is correct as written and I will approve on that evidence alone. If it reports T minus 2, there is a smaller fix that also removes the mismatch while leaving the encoder's input convention untouched, described inline.

One observation that is outside this diff and should not hold the PR. The default_feature_len of 250 in the prompt expansion path fabricates a placeholder count unrelated to any real audio whenever feature_lens comes up short, so it remains an independent source of exactly this class of crash. Worth removing separately.

Validation: static review of the full patch and both production consumers, the STFT framing that fixes the minimum mel length at 3 for non-empty audio and 2 for empty, the encoder's mask construction, and an exhaustive comparison of old against new placeholder counts across mel lengths. A cross-model panel reviewed my first pass and correctly caught that I had compared the length formula at a fixed input while the input to it had also changed; the counts above are recomputed and verified. GitHub checks pass, and per-PR Buildkite is green on general, Intel, and NPU at this head. No fork code was executed.
